File: sample_5/zadanie_2/zadanie_2.py
class TreeNode:
    """
    Класс для узла дерева
    """
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class ExpressionTree:
    """
    Класс для построения и вычисления синтаксического дерева арифметического выражения
    """
    def __init__(self):
        self.operators = {'+', '-', '*', '/'}

    def is_operator(self, char):
        return char in self.operators

    def build_tree(self, postfix):
        """
        Построение дерева из постфиксной записи
        :param postfix: Список символов в постфиксной нотации
        :return: Корень дерева
        """
        stack = []

        for char in postfix:
            node = TreeNode(char)

            if self.is_operator(char):
                # Достаём два операнда из стека
                node.right = stack.pop()
                node.left = stack.pop()

            stack.append(node)

        return stack.pop()

    def evaluate(self, root):
        """
        Вычисление значения выражения, представленного деревом
        :param root: Корень дерева
        :return: Результат вычисления
        """
        if root is None:
            return 0

        if not self.is_operator(root.value):
            return int(root.value)

        left_val = self.evaluate(root.left)
        right_val = self.evaluate(root.right)

        if root.value == '+':
            return left_val + right_val
        elif root.value == '-':
            return left_val - right_val
        elif root.value == '*':
            return left_val * right_val
        elif root.value == '/':
            return left_val / right_val

# Пример использования
if __name__ == "__main__":
    expression = "3 4 + 2 * 7 /"  # Постфиксная запись (эквивалентно: ((3 + 4) * 2) / 7)
    postfix = expression.split()

    tree = ExpressionTree()
    root = tree.build_tree(postfix)

    print("Результат вычисления выражения:", tree.evaluate(root))

File: sample_5/zadanie_2/zadanie_2.cpp
/*
    Тема №2. Абстрактные типы данных (нелинейные структуры) – 4 часа
    
    АТД «дерево» - лексический разбор арифметического выражения
*/

#include <iostream>
#include <stack>
#include <string>
#include <sstream>
#include <cctype>
#include <vector>

using namespace std;

// Узел дерева
struct TreeNode {
    string value;
    TreeNode* left;
    TreeNode* right;

    TreeNode(string val) : value(val), left(nullptr), right(nullptr) {}
};

class ExpressionTree {
public:
    // Построение дерева из постфиксной записи
    TreeNode* buildTree(const vector<string>& postfix) {
        stack<TreeNode*> st;

        for (const string& token : postfix) {
            TreeNode* node = new TreeNode(token);

            if (isOperator(token)) {
                // Достаем два операнда из стека
                node->right = st.top(); st.pop();
                node->left = st.top(); st.pop();
            }

            st.push(node);
        }

        return st.top();
    }

    // Вычисление значения выражения
    double evaluate(TreeNode* root) {
        if (!root) return 0;

        if (!isOperator(root->value)) {
            return stod(root->value);
        }

        double leftVal = evaluate(root->left);
        double rightVal = evaluate(root->right);

        if (root->value == "+") return leftVal + rightVal;
        if (root->value == "-") return leftVal - rightVal;
        if (root->value == "*") return leftVal * rightVal;
        if (root->value == "/") return leftVal / rightVal;

        return 0;
    }

private:
    // Проверка, является ли строка оператором
    bool isOperator(const string& token) {
        return token == "+" || token == "-" || token == "*" || token == "/";
    }
};

int main() {
    string expression = "3 4 + 2 * 7 /"; // Постфиксная запись
    vector<string> postfix;
    stringstream ss(expression);
    string token;

    while (ss >> token) {
        postfix.push_back(token);
    }

    ExpressionTree tree;
    TreeNode* root = tree.buildTree(postfix);

    cout << "Результат вычисления выражения: " << tree.evaluate(root) << endl;

    return 0;
}
